src/leap/bitmask/core/mail_services.py

```diff
@@ -0,0 +1,668 @@
+"""
+Mail services.
+
+This is quite moving work still.
+This should be moved to the different packages when it stabilizes.
+"""
+import json
+import os
+from glob import glob
+from collections import defaultdict
+from collections import namedtuple
+
+from twisted.application import service
+from twisted.internet import defer
+from twisted.python import log
+
+from leap.bonafide import config
+from leap.common.service_hooks import HookableService
+from leap.keymanager import KeyManager, openpgp
+from leap.keymanager.errors import KeyNotFound
+from leap.soledad.client.api import Soledad
+from leap.mail.constants import INBOX_NAME
+from leap.mail.mail import Account
+from leap.mail.imap.service import imap
+from leap.mail.incoming.service import IncomingMail, INCOMING_CHECK_PERIOD
+from leap.mail import smtp
+
+from leap.bitmask.core.uuid_map import UserMap
+
+
+class Container(object):
+
+    def __init__(self):
+        self._instances = defaultdict(None)
+
+    def get_instance(self, key):
+        return self._instances.get(key, None)
+
+
+class ImproperlyConfigured(Exception):
+    pass
+
+
+def get_all_soledad_uuids():
+    return [os.path.split(p)[-1].split('.db')[0] for p in
+            glob(os.path.expanduser('~/.config/leap/soledad/*.db'))]
+    # FIXME do not hardcode basedir
+
+
+class SoledadContainer(Container):
+
+    def __init__(self, basedir='~/.config/leap'):
+        # FIXME do not hardcode basedir
+        self._basedir = os.path.expanduser(basedir)
+        self._usermap = UserMap()
+        super(SoledadContainer, self).__init__()
+
+    def add_instance(self, userid, passphrase, uuid=None, token=None):
+
+        if not uuid:
+            bootstrapped_uuid = self._usermap.lookup_uuid(userid, passphrase)
+            uuid = bootstrapped_uuid
+            if not uuid:
+                return
+        else:
+            self._usermap.add(userid, uuid, passphrase)
+
+        user, provider = userid.split('@')
+
+        soledad_path = os.path.join(self._basedir, 'soledad')
+        soledad_url = _get_soledad_uri(self._basedir, provider)
+        cert_path = _get_ca_cert_path(self._basedir, provider)
+
+        soledad = self._create_soledad_instance(
+            uuid, passphrase, soledad_path, soledad_url,
+            cert_path, token)
+
+        self._instances[userid] = soledad
+
+        data = {'user': userid, 'uuid': uuid, 'token': token,
+                'soledad': soledad}
+        self.service.trigger_hook('on_new_soledad_instance', **data)
+
+    def _create_soledad_instance(self, uuid, passphrase, basedir, server_url,
+                                 cert_file, token):
```

Comment:
Could we come out with a different name than `basedir`? There is a `self._basedir` that has a different meaning in this class, it's confusing.